### PR TITLE
Fix #438 - check for __gshared and immutable redundancies

### LINKED ISSUE #438: Check for __gshared and immutable are redundancies
See https://github.com/dlang/phobos/pull/5471

### PR DESCRIPTION
**Pull request by wilzbach**
Fix https://github.com/dlang-community/D-Scanner/issues/438

### CODE REVIEW

**Comment by BBasile** [comment]
Self-linting fails because of this new check but this case if valid:

```d
class C
{
    static immutable foo = "";
}

void main()
{
    enum cfoo = C.foo;
}
```

without `static` i need a C instance and without `immutable` i cant assign to a `enum`.
static immutable must be accepted, at least in an aggregate.

**Comment by wilzbach** [comment]
Okay I back-paddled here a bit and made the check more conservative and only shout for the obvious mistakes for now.

Here's what I get from running over Phobos:

```
std/datetime/timezone.d(2820:43)[warn]: Variable declaration for `_tzIndex` has redundant attributes (`static`, `__gshared`).
std/parallelism.d(1085:29)[warn]: Variable declaration for `nextInstanceIndex` has redundant attributes (`__gshared`, `static`).
std/experimental/logger/core.d(1631:94)[warn]: Variable declaration for `_buffer` has redundant attributes (`static`, `__gshared`).
std/experimental/allocator/building_blocks/affix_allocator.d(365:45)[warn]: Variable declaration for `instance` has redundant attributes (`static`, `__gshared`).
std/experimental/allocator/building_blocks/segregator.d(274:41)[warn]: Variable declaration for `instance` has redundant attributes (`static`, `__gshared`).
std/experimental/allocator/building_blocks/quantizer.d(56:36)[warn]: Variable declaration for `instance` has redundant attributes (`static`, `__gshared`).
std/experimental/allocator/package.d(2535:35)[warn]: Variable declaration for `state` has redundant attributes (`static`, `__gshared`).
std/experimental/allocator/package.d(2536:39)[warn]: Variable declaration for `result` has redundant attributes (`static`, `__gshared`).
std/concurrency.d(2525:42)[warn]: Variable declaration for `inst` has redundant attributes (`static`, `__gshared`).
std/concurrency.d(2539:42)[warn]: Variable declaration for `inst` has redundant attributes (`static`, `__gshared`).
std/concurrency.d(2546:33)[warn]: Variable declaration for `cnt` has redundant attributes (`static`, `__gshared`).
```

Looks all legit too me :)

File: src/dscanner/analysis/redundant_storage_class.d [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		string[] globalAttributes;
+		foreach (attrib; attributes)
+		{
+			if (attrib.attribute.type.among(tok!"shared", tok!"static", tok!"__gshared", tok!"immutable"))
+				globalAttributes ~= attrib.attribute.type.str;
+		}
+		if (globalAttributes.length > 1)
+		{
+			if (globalAttributes.length == 2 &&
+					globalAttributes.all!(a => a.among("shared", "static")) ||
```
Comment (BBasile) [inline src/dscanner/analysis/redundant_storage_class.d]: D-Scanner warning is right here, you should put line 55 and 56 between parens

Comment (wilzbach) [reply src/dscanner/analysis/redundant_storage_class.d]: -> https://wiki.dlang.org/Operator_precedence

but I used parens here now for clarity and to satisfy the self-linting.

**Pull request merged**